[Migrated to PR #644] Gyro Aiming support

AL2009man opened the pull request:
migrated to https://github.com/fgsfdsfgs/perfect_dark/pull/644

AL2009man committed:
added GameControllerHasSensor codeset

AL2009man committed:
added SensorDataRate

AL2009man committed:
added rudimentry gyro aim code

AL2009man committed:
added more gyro code

AL2009man committed:
added additional gyro aim variables with camera controls

AL2009man committed:
added Mouse sharing with Gyro Code

AL2009man committed:
remove gyroaim leftovers

AL2009man committed:
added Gyro Aiming Menu

While all sliders and toggles are not functionable: it'll be the bridge.

AL2009man committed:
move inputUpdate declaration

AL2009man committed:
changed default sensitivity to 2.50

given the value is from 0.01 to 10.00, I expect it'll be similar

AL2009man committed:
attempts to reorder declaration

AL2009man committed:
slight gyro cleanup.

AL2009man committed:
added Accelerometer sensors to the list

AL2009man committed:
another slight revised change

AL2009man committed:
enabled SDL_SetHint

For some reason: this sourceport isn't using it. this is very crucial inorder to add Gyro Aiming support

AL2009man committed:
added Gyro Aim modes and Crosshair Speed

This adds Gyro Aim modes where you can choose between Camera, Crosshair or Camera+Crosshair (the default), and both Crosshair Speeds are added.

the Gyro Menu was also reorganized.

AL2009man committed:
added gyro-related code onto bondmove.c

AL2009man committed:
added missing gyro code on ini file

AL2009man committed:
Restored crosshair axis

AL2009man committed:
fixed Forward Declaration

AL2009man committed:
attempt to re-add bondmove.c codebase

AL2009man committed:
attempt to add Gyro Modifier action

Gyro Modifier is connected to "Gyro Activation" button, which will cover Gyro Toggle and Gyro Hold (and the reverse version).

AL2009man committed:
GyroEnabled code hotfix

at somepoint: a portion of the gyro code got broken. this has been restored alongside

AL2009man committed:
Gyro crosshair fix

Crosshair Aiming for Gyro has been fixed. Right now: Gyro Axis is broken-- but that will be fixed later

AL2009man committed:
fixed GyroEnable button for Crosshair modes

the button is now functionable.

AL2009man committed:
fixed Crosshair behavior when using Gyro Aim

Rebuild the entire gyro-based codebase inorder to solve a specific problem with mouse-related settings

AL2009man committed:
reorganized freelook and crosshair mode

AL2009man committed:
fixed Gyro Axis Roll

after hours, it turns out that handleSensorUpdate was the culprit to the incorrect gyro axis.

AL2009man committed:
Merge branch 'GyroAim-WIP' into port

AL2009man committed:
Merge pull request #2 from AL2009man/port

backport changes over to GyroAim-WIP

AL2009man committed:
Revert "port: only enable SDL hints on SDL versions that support them"

This reverts commit e32d719c16cbd487d460aab0944b1548c8ee3ed6.

AL2009man committed:
fully decoupled Gyro Sens and Gyro Sens scale

both sliders will now work as intended, but there's a built-in damp for no reason.

AL2009man committed:
fix recorrection

AL2009man committed:
Merge remote-tracking branch 'origin/port' into GyroAim-WIP

AL2009man committed:
Merge branch 'fgsfdsfgs:port' into Gyro-Aim

AL2009man committed:
Reworked Gyro Aim implementation

the entire implementation has been rebuild from scratch based on the past commits as a basis.

AL2009man committed:
Merge branch 'Gyro-Aim' of https://github.com/AL2009man/perfect_dark into Gyro-Aim

AL2009man commented:
Significant progress has been made on the current gyro aiming implementation. 

Based on learning the codebase, I opted to rebuild the whole thing by scratch. Everything's far more organized and more future-proof for further Gyro-related enhancements. Right now: the initial gyro aiming is playable...but there are issues I might need assistance with, especially with the mouse delta scaling.

- The Sensitivity Scale seems to be tied to the frame rate.
- The Gyro equivalent of Crosshair Speed is not working as intended. Right now; it's tied to the main Sensitivity. 
- Gyro Aim mode (that will cover Camera only, Crosshair only or Both) does not work.
- The internal Gyro's DeltaScale doesn't play nicely with Natural Sensitivity Scale. 

the last two parts is the one where I'm having a tougher time making it work. my end goal is to ultimately make `1.00` to be a closer equivalent to a 1:1 real-life camera turn relative to the controller's motion sensors, but I haven't achieve it yet. 

Closesly I've ever gotten to matching a full 360 degree camera turn is `4.0f`, but if you increase gyro sens to "2.50" (or 2.5x in Steam Input Gyro Degrees Sens terms), the gyro sens is quadrupled. I've spent the bulk of my time trying to fix it, even double-checking and removing anything (but Delta Scale) that affects scaling, but so far: no dice.

same case with "Gyro Aim modes": still having issues trying to get "Crosshair only" and "Camera+Crosshair" to only activates the Aim portion, alongside ensuring the it's own sens is separate.

Everything else: you can try out the the Gyro-Aim branch right now to see what the current gyro aim implementation looks like.

AL2009man committed:
Natural Sensitivity Scale and partial Gyro Aiming mode fix

The Gyro sensitivity slider is now based upon Natural Sensitivity Scale system from JoyShockMapper, Steam Input and Fortnite.

This commit also includes:

- partial fix for Gyro's Aim Modes.
- missing .ini codes are now added. it'll be saved properly.
- Gyro Sensitivity is now defaulted at 2.50, mirroring Steam Input's Gyro default sens
- Improved Gyro Movement threshold

AL2009man commented:
Been busy, mostly on getting the Natural Sensitivity Scale.

long-story short: the overall sens scale is now based on the movement of the gyro. 1.00 is relative to the controller's real live movement. Joanna Dark moving at one single 360-degree turn will be relative to the controller. Setting the slider to "2.50" will double it. this is a modifier, and is a strong recommendation when implementing Gyro Aim to a game.

That being said, the way it was done is a bit of a hacky workaround (that includes Movement Threshold!) that has resulted in countless game crashes just to ensure it isn't tied to the frame rate. There's still key issues that I'll need to fix.

in the meantime: 

https://github.com/user-attachments/assets/f23633d5-8816-4f70-acbe-dcc3571f0c91

AL2009man committed:
Gyro AutoCalibration, Gyro Aiming mode fix, Gyro Threshold Fix, partial fix on game crash

This commit introduces Gyro Auto-Calibration support, enabled by default. However: the Auto-Calibration only kicks in when the controller is placed on the ground. Gyro Threshold received minor improvements, especially with Nintendo Controllers.

the rest:
- Gyro Aim modes now works as intended!
- Gyro's Crosshair Sens is fully decoupled!
- fixed Controller Hotplug while disconnecting controller
- Preparations for Gyro Space implementation

AL2009man committed:
Gyro Space code and Sensitiviity Scale fix

This commmit properly introduces Gyro Space support (Local Space, Player Space, World Space) to handle Orientation. the default is still Yaw axis. powered by AL2009man's upcoming "GyroSpace-and-Play" header (WIP)

Additionally: Gyro Sensitivity scale has been fully fixed. it should lean a lot closer to the original mouse code.

AL2009man committed:
Revised Gyro Threshold and Gyro Auto-calibration

Gyro Threshold has been revised to better be more raw, while adjusted deadzone for Nintendo Switch controlelrs only. Additionally: Gyro Auto-Calibration is now disabled by default for the time being.

AL2009man committed:
fixed Motion Sensor input and reworked Gyro Sensitivitiy Scale

This overhauls the entirely of Motion Sensor system to hopefully address constant game crashes during coding. This also fixes controller hotpluggin issue when Gyro Input is enabled and added missing gyro axis.

additionally: the Gyro Sensitivity scale has been overhauled to now mirror id Tech's mouse sensitivity scale as a basis.

AL2009man committed:
updated Local Space code and fix i686-linux build

hopefully, the i686-linux folks can test it on GitHub Action. Additionally: GyroSpace.h now covers version numbers.

AL2009man committed:
Revert "updated Local Space code and fix i686-linux build"

This reverts commit 8fd2474821a0fc350e8f4c7dee90c06c1cc95ff0.

AL2009man committed:
Enhanced gyro input handling and GyroSpace 0.6.0

- Adjusted gyro sensitivity thresholds and refined activation modes in `input.c`.
- Improved controller input handling with enhanced logging and state management.
- Enhanced gyro auto-calibration with stationary checks.
- Refined deadzone handling for gyro inputs.  Now uses Soft Tiered Smoothing.
- Improved code structure and readability with consistent formatting and comments.
- Gyro Modifer action can be assigned, but does not work properly.

AL2009man commented:
I'm halfway done with the implementation, hit some strives. However: there are issues that still needs to be addressed:

- after compiling: Game Crash or Camera spinning erratically can occur. If that happens: recompile everything by scratch
- Gyro Calibration do not work. 
- Nintendo Switch gamepad's drift issues, turns out: common issue typically solved with Calibration system.
- Gyro Modifier action is broken at the moment, as it's meant to tied to Gyro Activation.

as for Gyro Calibration: I'll need to find a way to create a menu pop-up in a similar way as changing the bindings within Key Binding section. Also needing to fix the Gyro Modifier function. 

I'll definitely need assistance on these two

fgsfdsfgs commented:
Looking at the code, I feel like this is way too overcomplicated, especially with the gyrospace lib in there. I've done simple gyro aiming in other ports I've worked on and it seemed to work well enough for users. I should probably look into backporting it to this one.

fgsfdsfgs commented:
Don't worry, I'm not going to commit it directly into `port` and screw you over, was thinking of making a separate branch.

AL2009man committed:
tweaked AutoCalibration, whitespace fix

AL2009man committed:
Merge branch 'port' into Gyro-Aim

AL2009man committed:
change cameramanualaim's allowgcross back

AL2009man committed:
updated `gyrospace.h` to version 0.6.5

this one is...a bit closer to what version 1.0.0 will be. Going forward: gyrospace.h code will remained untouched until Gyro Space gets closer to completion (unless there's last minute bugs to fix)

AL2009man committed:
Cleaned up `inputhandleGyroController` and Mouse Sens regression fix

To make gyro coding a tiny bit easier; the `inputHandleGyroController` portion has been organized for easier understanding.

Mouse sens scale was accidentally reverted back to pre-Quake levels, but this should be changed back to normal. To quote Gianni: "Lower your sensitivity, king"

AL2009man committed:
temporaily remove Gyro Calibration

for the time being: Auto-Calibration will be removed until a proper, more functional way of Gyro Calibration can be implemented. this might be reserved for the future PR.

AL2009man committed:
Merge branch 'fgsfdsfgs:port' into Gyro-Aim

AL2009man commented:
> Don't worry, I'm not going to commit it directly into `port` and screw you over, was thinking of making a separate branch.

ok @fgsfdsfgs, I've been making preparations for the "separate branch" idea, first I made 3 changes

- Gyro Calibration is temporarily removed until a proper system is in place.
- Cleaned up `inputhandleGyroController`, to make it easier to read and understand said code
- Whitespace fix

only thing that hasn't been truly solve is the game crash and/or camera spinning after recompiling. I haven't figure that one out, but hopefully: a separate branch will help

AL2009man committed:
updated GyroSpace header to 0.6.7

this removes the internal scaling infavor of a more rawer method.

AL2009man committed:
Merge branch 'Gyro-Aim' of https://github.com/AL2009man/perfect_dark into Gyro-Aim

AL2009man committed:
updated LICENSE to include GyroSpace and Play header

lthough: the layout itself is super placeholdery and might need to be revised at a later date.

AL2009man committed:
Revert "updated LICENSE to include GyroSpace and Play header"

This reverts commit d45e30750e8e4a56f0fea63013ae7406b2052c1b.

AL2009man committed:
attempt to fix game crash after compiling and linux build GitHub Action

This is a attempt to fix game crash after compiling step, while attempting to fix GitHub Action build for Linux

AL2009man committed:
Nintendo Deadzone temporarily removed and updated GyroSpace header to 0.8.5

Nintendo-specific deadzone has been removed for the time being until a proper calibration setup is implemented. Nintendo Switch controllers will have to increase the Gyro Movement Threshold in the meantime.

Additionally: GyroSpace and Play header has been updated to 0.8.5.

AL2009man commented:
Finalizing my current Gyro Space code to be near completion, and I think this PR is ready to be used **as a completely separate branch**. Further work can resume there to avoid any conflicts, as suggested by @fgsfdsfgs.

For now: I recently made at least two changes as an attempt to fix the post-compile crash/camera-spin/ issue, while temporarily removing the Nintendo-specific deadzone workaround until a proper Gyro Calibration menu section gets added in.

AL2009man committed:
updated GyroSpace to 0.8.6

AL2009man committed:
Gyro Actvation works as intended

AL2009man committed:
fixed Gyro Activation for splitscreen

AL2009man committed:
Attempt to fix build compile issue

This fix adds a safety net when it come to gyro deltas while swapping to float method. hoepfully, it won't cause a game crash after code compile.

AL2009man committed:
restored fakeController

after potentially fixing the crash problem, we don't need a workaround on the fakecontroller code. so, it'll be restored.

AL2009man commented:
> 
> only thing that hasn't been truly solving is the game crash and/or camera spinning (exploding vertices can occur, but that's rare) after recompiling. I haven't solve that issue yet, but a separate branch will be needed for now.

I...think I managed to get it fixed? I tried compiling the build, and grabbed GitHub Actions and so far: I haven't encountered any game crashes after a compile.

AL2009man committed:
renamed `GyroActivationMode` to `GyroModifier`, GyroModifer now defaults to "Disabled while Held"

Gyro's "Aim Mode" and "Gyro Acttivation" might've been too confusing. To better differenciate: Gyro Activation has been renamed to "Gyro Modifer".

Gyro Modifier now defaults to "Disabled while Held", which disables Gyro Camera/Crosshair while holding the Gyro Modifier action (Right Joystick Click by default).

AL2009man committed:
remove typo on the "d" into "Disabled while Held"

AL2009man committed:
added Gyro Camera/Crosshair Invert option

You can now invert both X and Y Axis of the Camera & Crosshair!

AL2009man committed:
fixed `enum gyroactivation` after a rebrand

AL2009man committed:
Added Vertical/Horizontal Output Mixer for Gyro

since our general Gyro Sens is by Natural Sensitivity Scale values (except for Crosshair portion), we're taking a different approach by adding V/H (in this case: X/Y) Output Mixer setup. This affects both Gyro Camera and Crosshair.

Moving the slider to the left reduces vertical sensitiviity (which is X), Moving the slider to the right horizontal sensitiviity (which is Y). theoretically: you can set the slider to -43% and have the Gyro Ratio to be 16:9 Gyro ratio. By default: 0% is 1:1 Ratio.

AL2009man committed:
increase Gyro Sensitivity to 30

This one is meant for hardcore gyro users. main Gyro Sensitivity will be increaesd from 10 to 30.

AL2009man committed:
fixed `GyroSpeedX/Y` ConfigInit edit

AL2009man committed:
removing `inputGyroGetAbsScaledDelta`

this one is...a relic of mouse code. we're gonna get rid of it

AL2009man committed:
tweaked overall GyroDeltaScale for both Gyro Camera and Crosshair

based on the changes from the other PR, this one will simplify the scaling method while the crosshair portion will be much closer to 0.022.

AL2009man commented:
Got some time to add new features and minor improvements under the hood while adding/fixing other stuffs in between.

![image](https://github.com/user-attachments/assets/2f343286-47c4-480f-9a09-58704b7eac16)

long-story short, two new stuffs, one new cap and one slight revision was added in: 

- Gyro scaling has been cleaned up to make it easier to read. this kinda helps me prepare for potential features like Flick Stick (might require overalling CameraDelta scaling for easier management)
- Gyro Camera's sensitivity cap has been increased to 30.00. Crosshair portion remains the same.
- Vertical/Horizontal Output Mixer (internally, but we'll call it "Gyro X/Y Output Mixer") has been added. letting you tweak the Gyro's ratio between Horizontal and Vertical Movement. This affects both Camera and Crosshair. 
- Gyro's Invert X/Y has been added

AL2009man committed:
added "Gyro" name on the newly introduced Invert X/Y Toggles

AL2009man committed:
Merge branch 'port' into Gyro-Aim

AL2009man committed:
introduced universal Camera Angle system

To streamlined the entire Mouse/Gyro Camera system while preparing for future camera features (like Flick Stick): we introduced a Camera Angle setup.

AL2009man committed:
tweaked Gyro Crosshair Sens scale

This should closely mirror Gyro Camera's sens scaling, which means the new default is now 2.50

AL2009man committed:
slightly revised notes on scaling factor

AL2009man committed:
reverted Universal Camera scaling, added additional notes for the define notes

due to issues within the scaling system, the original setup has been restored.
While various notes has been added to explain each value.

AL2009man commented:
Been...finalizing most of the Gyro Aim work in between. Right now: im a bit focused on fixing splitscreen multiplayer support for Gyro Input, I mostly replaced a lot of the global variables with a per-controller setup...but i got myself stuck on trying to fix the bug myself. I will need some bit of help. 

To avoid conflicts, [I created a branch separate](https://github.com/AL2009man/perfect_dark/tree/GyroAim-Multiplayer) from this PR.

AL2009man committed:
Updated GyroSpace header to 0.9.0

AL2009man committed:
Separated the Soft Tiered Smoothing and Deadzone into two sliders

origianlly: Gyro Movement Threshold slider includes Soft Tiered Smoothing, but it make more sense to separate the slider into two unique ones.

AL2009man committed:
reorganized Gyro Settings menu

The Gyro Settings has been reorganized, while there's now a separate line for each "sub-categories"

AL2009man committed:
moved <math.h> to the top, while renaming `GyroTighting` Management to Smoothing, to better reflect the name

AL2009man committed:
fixed whitespace + backport recent fixes from port branch

AL2009man committed:
Merge branch 'fgsfdsfgs:port' into Gyro-Aim

AL2009man committed:
initial support for Gyro Aiming on Splitscreen Multiplayer

Revised overall gyro code to be based on a per-controller basis, allowing you to have a local gyro multiplayer session.  However: the Crosshair Aiming portion for Player 2/3/4 does not work at the moment.

this also improves the crosshair edge turning for gyro, making it a bit consistent with mouse input's

AL2009man committed:
update GyroSpace to 0.9.1

AL2009man committed:
GamepadMotionHelper used to handle Gyro Camera logic.

This commit replaces half of the gyro handling with GamepadMotionHelper (by Jibb Smart). This allows us to prepare for Gyro Calibration support.

Additionally: GyroSpace.h header will be replaced (as this one will require a complete clean slate, it may come back at some point), while Local Space portion has been modify to be more raw about it.

AL2009man committed:
initial Gyro Calibration submenu

Gyro Calibration is supported, but for testing reasons: you'd need to press F10 to start gyro calibration, and end it by Press F11. Gyro Calibration menu does not work at the moment.

AL2009man committed:
Replaced Gyro Calibration menu infavor of a Auto-Calibration setup.

AutoCalibration system will be powered by GameapdMotionHelper! We won't be needing a dedicated Gyro Calibration menu for this one.

also: the maual Gyro Calibration game action now works, and will be defauled to Misc button (which would be Nintendo Switch's Capture Button and DualSense's Microphone button)

AL2009man committed:
add .vscode to the gitignore

AL2009man committed:
improved Auto-Calibration and reorganized overall gyro code

AL2009man committed:
Merge branch 'fgsfdsfgs:port' into GyroAimCalibration

AL2009man committed:
Migrating to GamepadMotionHelper, added Gyro Calibration

This commit migrates most of the existing Gyro code in favor of GamepadMotionHelpers [(specifically: the Valkirie fork as we can use .cpp, reducing the need for an extra file for C Wrapper](https://github.com/Valkirie/GamepadMotionHelpers/)), allowing us to build a painless form of Gyro Calibration.

Going forward: you can either press the Misc button (which is DualSense's Mic button, Nintendo Switch 1/2's Capture Button) to manually calibrate on the fly, or rely on Gyro's Auto-Calibration (enabled by default). This finally solves a major issue with Nintendo Switch 1's Gyro drifting issue.

Additionally, Axis Orientation went through a revision, replacing GyroSpace.h in favor of GMH's Player Space/World Space systems, while changing Local Space to be more raw about it. This should fix a major issue with Gyro Space's accelerometer not playing nicely with the built-in Sensor Fusion method. (one of the major reasons why we're switching to GamepadMotionHelper in the first place, key issues were found during testing on a separate repository)

Depending on feedback: further changes/improvements will continue in future commits.

AL2009man committed:
GitHub Action attempt fix

looks like something went wrong, so we're gonna swap out GamepadMotion.cpp with the .hpp one instead until the GitHub Action fix gets sorted out.

AL2009man committed:
fixed regression on Gyro Drift after Controller disconnect

AL2009man committed:
slightly improve AutoCalibration code

AL2009man committed:
Revert "slightly improve AutoCalibration code"

This reverts commit 8785f40574e3ef217d0e3569c0aa5f3932895e1b.

AL2009man committed:
simplify Gyro Calibration code for a time being

AL2009man committed:
Fixed regression on World Space axis when using Gyro Calibration, Restored GamepadMotion.cpp

Long-story short: `accelMagnitude` was the culprit to many source of issues. This has been changed to be numers. Gyro Calibration should work as intended, however: Gyro AutoCalibration will need more work

AL2009man committed:
fixed Gyro AutoCalibration toggle on `CONTROLLERCFG_DEFAULT`

AL2009man committed:
slightly fixed Gyro AutoCalibration

Gyro Auto-Calibrate has been officially added, just place the controller on a surface (or table) and gyro will attempt to auto-calibrate the drift. For now: this is disabled by default, but Nintendo Switch controllers will need it the most.

AL2009man committed:
attempted GitHub Action for MacOS fix

AL2009man committed:
Reworked Gyro Autocalibration

completely simplify the process of Gyro Autocalibration, it should behave more consistently whiel adhering towards GyroWiki standards. However, Gyro Autocalibration is still disabled by default, Nintendo Switch Pro controller users still need to press the Capture button if they want their controller to be calibrated. This will be that way UNTIL Gyro Autocalibration is battle tested

AL2009man committed:
Gyro Smoothing now works as intended.

AL2009man committed:
improved Gyro Movement and Smoothing slider

Sliders for Gyro Movement Threshold and Smoothing has been revised. thet Movement Threshold will feel smoother than ever before...alongside the smoothness.

 `gyroSmoothing` will be defaulted to 0.07

AL2009man committed:
Added Recenter Camera/Crosshair action binding

Recenter to Horizon Camera, or Reset Camera is a new game action that will recenter the camera vertically, This will be similar for those who play Splatoon. By default: you can press either the Right Stick or C Key to recenter the Camera.

the Gyro Aiming default has been changed once again, now Gyro Aim will be enabled when "Aiming"/Crosshair.

Additionally: "Gyro Calibration (Manual)" can be used on keyboard inputs (default is F10 at the moment), in case your controller don't feature a Misc key.

AL2009man committed:
Viewmodel Sway is now shared between Gyro and Mouse Input

AL2009man committed:
fixed Crosshair forced recentering on Local Multiplayer

turns out: Gyro Crosshair was being clamped, which doesn't affect Player 1 but affect other players. This removes the clamping.

AL2009man committed:
attempted to fix Gyro Crosshair framerate scaling

It looks like Gyro Crosshair (and Mouse Crosshair) is tied to Framerate, this attempts to solve it, but it won't fix it 100 percent.

also restored Gyro Drift after disconnect

AL2009man committed:
slightly improved Gyro Smoothing

AL2009man committed:
added `SDL_GRAVITY_STANDARD` and upgraded GamepadMotionHelper header

AL2009man committed:
changed overall Gyro Scaling

recalibrated Natural Sensitiviy Scale to be a bit closer to 1:1 scale. (although: Switch Pro Controller doesn't feel right but you might wanna wait for SDL2 hotfix

AL2009man committed:
cleaned up and rename GamepadMotionHelperWrapper to simply GamepadMotionWrapper

to prepare for a fork repository for GMH: certain stuffs has been renamed for conveniences

AL2009man committed:
added Gyro's DeltaTime

This adds DeltaTime for Motion Sensors to improve gyro handling. this should cover everything we'll need for GamepadMotionHelper. (also properly rename GamepadMotionWrapper.h to GamepadMotion.h)

Also, some minor documentation improvements and leftover codes.

AL2009man committed:
moved GamepadMotionHelper to it's own file.

inorder to be more consistent with other dependencies: GamepadMotion.cpp and GamepadMotion.h are moved to it's own dedicated file.

AL2009man committed:
updated README.md to include Gyro Aiming information

to prepare for the finalization stages: the README.md will mention Gyro Aiming.

also renamed "L3" to "LS Click" because we're using Xbox naming terminology, not PlayStation/Steam Deck!

AL2009man committed:
added missing CLICK on Reset Camera / Crosshair

AL2009man committed:
slightly improved README.md and crediting GamepadMotionHelper authors on the Credits section

AL2009man committed:
merge `insightaimmode` into one for `allowmcross` and `allowgcross`

AL2009man commented:
note: this PR is now ready for review, but additional testings will be needed, before merger. (https://github.com/fgsfdsfgs/perfect_dark/pull/591#issuecomment-2973428026)

also: if you plan to approve the merge (if it's either a "port" branch or a separate branch: it is HIGHLY recommended to squash the PR to avoid commit history junk.

AL2009man committed:
cleaned up redundancies, last remaining whitespaces

last minute clean ups

AL2009man committed:
removed vscode leftover

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -102,6 +102,9 @@ enum contkey {
 	CK_CANCEL,
+	CK_RESET_CAMERA,
+	CK_GYRO_MOD,
+	CK_GYRO_CALIB,
```
Adding new buttons to this enum will break the top buttons. It covers exactly 32 bits and should not be changed.
Instead use/rename the numbered entries which are currently not in use (`CK_0040` - `CK_0800`).

Also, do we really need a separate calibration button? This should probably be a menu thing.

AL2009man replied:
as for a separate gyro calibration button. it's necessary in case a user needs to quickly calibrate the controller mid-game. consider it a emergency button. In fact: JoyShockMapper and Haste's Hasty Controls mod also features the same binding.

at the same time: there was plans for a Gyro Calibration UI, but i couldn't get it working. I believe this is due to how it interacts with `inputUpdate`. I tried again at least two times but couldn't figure it out. 

it's possible to get it working, but I might need help on that department...but i think it's better to reserve that for a separate PR. In the meantime: users can rely on "Auto-Calibration Gyro while Stationary" toggle.

AL2009man replied:
Gyro Calibration UI has been added in, but the button-based Gyro Calibration will be kept as is.

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
+// Gyro FreeLook Sensitivity Management
+void inputGyroGetSpeed(s32 cidx, f32* x, f32* y);
+void inputGyroSetSpeed(s32 cidx, f32 x, f32 y);
+f32 inputGyroGetSpeedX(s32 cidx);
+void inputGyroSetSpeedX(s32 cidx, f32 x);
+f32 inputGyroGetSpeedY(s32 cidx);
+void inputGyroSetSpeedY(s32 cidx, f32 y);
```
What's the point of separate functions for X and Y if you already have `GetSpeed`? Just do it the same way as `MouseGetSpeed`, the gyro API is already massive enough as it is.

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
+void inputGyroSetAimSpeedX(s32 cidx, f32 x);
+f32 inputGyroGetAimSpeedY(s32 cidx);
+void inputGyroSetAimSpeedY(s32 cidx, f32 y);
+
```
Same deal as with `GyroSetSpeed`.

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
+f32 inputGyroGetCrosshairSpeedX(s32 cidx);
+void inputGyroSetCrosshairSpeedX(s32 cidx, f32 x);
+f32 inputGyroGetCrosshairSpeedY(s32 cidx);
+void inputGyroSetCrosshairSpeedY(s32 cidx, f32 y);
```
Same deal as with `GyroSetSpeed`.

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
+s32 inputGyroAimInvertXIsEnabled(s32 cidx);
+void inputGyroAimInvertXEnable(s32 cidx, s32 enabled);
+s32 inputGyroAimInvertYIsEnabled(s32 cidx);
+void inputGyroAimInvertYEnable(s32 cidx, s32 enabled);
```
Probably same deal as with `GyroSetSpeed`. Better do something like this:
```
void inputGyroGetInvert(s32 cidx, s32* out_invertx, s32* out_inverty);
void inputGyroSetInvert(s32 cidx, s32 invertx, s32 inverty);
```

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -196,6 +235,9 @@ const char *inputGetKeyName(s32 vk);
 s32 inputGetContKeyByName(const char *name);
 
+// Returns 1 if the reset camera action is pressed for the given controller indexAdd commentMore actions
+s32 inputResetCameraPressed(s32 idx);
```
This function has no purpose, it's just one line, and that line uses a constant that has a descriptive name (`CK_RESET_CAMERA`).

AL2009man replied:
tested it. all it does is that it breaks "Reset Camera/Crosshar" function entirely. the function only serves a purpose inorder to interact with these two:


[input.c:](https://github.com/AL2009man/perfect_dark/blob/8288950813e856e618ae747e4b62c9baca1a60b1/port/src/input.c#L1458) 
```
s32 inputResetCameraPressed(s32 idx)
{
	return inputBindPressed(idx, CK_0040);
}
```


[bondmove.c](https://github.com/AL2009man/perfect_dark/blob/8288950813e856e618ae747e4b62c9baca1a60b1/src/game/bondmove.c#L2033):

```
#ifndef PLATFORM_N64
			// Camera recenter vertical binding
			static bool prevresetpressed = false;
			bool resetpressed = inputResetCameraPressed(cidx);

			if (resetpressed && !prevresetpressed) {
				if (g_Vars.currentplayer) {
					if (g_Vars.currentplayer->insightaimmode) {
						// Recenter crosshair if in crosshair aim mode
						g_Vars.currentplayer->swivelpos[0] = 0.0f;
						g_Vars.currentplayer->swivelpos[1] = 0.0f;
					}

					if (offbike && !g_Vars.currentplayer->insightaimmode) {
						g_Vars.currentplayer->cachedlookahead = bmoveCalculateLookahead();
						g_Vars.currentplayer->docentreupdown = true;
						g_Vars.currentplayer->automovecentre = false;
						g_Vars.currentplayer->speedverta = (g_Vars.currentplayer->vv_verta < g_Vars.currentplayer->cachedlookahead) ? 7.0f : -7.0f;
						s_recenterfast = 1;
					}
				}
			}
			prevresetpressed = resetpressed;
#endif
```

but, if we seriously want it to happen, this might require refactoring it just to remove the function entirely.

AL2009man replied:
[this has been addressed](https://github.com/AL2009man/perfect_dark/commit/e8bb68250c90f24db058c4f97d5e744ae0987dae).

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -121,6 +124,42 @@ enum mouselockmode {
+} GyroCalibrationOp;
+
+typedef enum {
+		CALIBRATIONMODE_MANUAL = 0,
+		CALIBRATIONMODE_STILLNESS = 1,
+		CALIBRATIONMODE_SENSORFUSION = 2
```
Please fix whitespace.

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
+// Gyro Aim Mode Management
+s32 inputGetGyroAimMode(s32 cidx);
+void inputSetGyroAimMode(s32 cidx, s32 mode);
+void applyGyroAimMode(s32 cidx, f32* deltaX, f32* deltaY, f32* deltaZ);
```
Functions that are only used in `input.c` should be made `static` and not declared in `input.h`.

AL2009man replied:
the Gyro Input API is heavily reliant on `bondmove.c`'s camera code (naturally), i already tried making it static but no  avail.

that can't be done unless it requires a significant gyro refactoring.

fgsfdsfgs replied:
No, I meant `applyGyroAimMode` specifically. Not sure why it shows 3 functions in the snippet.

AL2009man replied:
please check, the rest of the "ApplyGyro[InsertProcessing]" has been made static.

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
+// Gyro Axis Mapping Management
+enum gyroaxismode inputGetGyroAxisMode(s32 cidx);
+void inputSetGyroAxisMode(s32 cidx, enum gyroaxismode mode);
+void applyGyroAxisMapping(s32 cidx, float gyroData[3], float accelData[3], f32* deltaX, f32* deltaY, f32* deltaZ);
```
Functions that are only used in `input.c` (`applyGyroAxisMapping`) should be made `static` and not declared in `input.h`.

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
+// Gyro Activation Mode Management
+s32 inputGetGyroModifier(s32 cidx);
+void inputSetGyroModifier(s32 cidx, s32 mode);
+void applyGyroModifier(f32* deltaX, f32* deltaY, f32* deltaZ, s32 activationMode, s32 idx);
```
Functions that are only used in input.c should be made static and not declared in input.h. (`applyGyroModifier`)

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
+// Gyro Movement Threshold Management
+f32 inputGetGyroMinThreshold(s32 cidx);
+void inputSetGyroMinThreshold(s32 cidx, f32 threshold);
+void applyGyroThreshold(f32* deltaX, f32* deltaY, f32* deltaZ, f32 threshold);
```
Functions that are only used in input.c should be made static and not declared in input.h. (`applyGyroThreshold`)

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
+// Gyro Smoothing Management
+f32 inputGetGyroSmoothing(s32 cidx);
+void inputSetGyroSmoothing(s32 cidx, f32 smoothing);
+void applyGyroSmoothing(f32* deltaX, f32* deltaY, f32* deltaZ, f32 smoothing, s32 cidx);
```
Functions that are only used in input.c should be made static and not declared in input.h. (`applyGyroSmoothing`)

fgsfdsfgs commented on port/include/input.h:
```diff
@@ -232,6 +274,100 @@ void inputMouseSetSpeed(f32 x, f32 y);
 
+// Gyro Controller Initialization & Management
+void inputHandleGyroController(s32 cidx);
+void closeGyroController(s32 cidx);
```
Seems to be a leftover with no implementation, or maybe this is supposed to be `inputCloseGyroController`? Either way, if this is unused outside of `input.c`, it should be made `static` and removed from the header.

AL2009man replied:
`inputCloseGyroController` is a workaround to a existing issue with Gyro not becoming active after a controller reconnect. 

also, Gyro API is tied to bondmove.c, as noted earlier.

AL2009man replied:
This issue has been fixed, will be rolling out later this week alongside Gyro Calibration reorganization

AL2009man replied:
This has been addressed. we now do things direrctly thru `inputInitController` and `inputCloseController` respectively.

fgsfdsfgs commented on port/src/input.c:
```diff
@@ -120,6 +190,9 @@ static const char *ckNames[CK_TOTAL_COUNT] = {
 	"STICK_YPOS",
 	"ACCEPT_BUTTON",
 	"CANCEL_BUTTON",
+    "CK_RESET_CAMERA",
```
Whitespace.

fgsfdsfgs commented on port/src/input.c:
```diff
@@ -346,6 +423,29 @@ static inline void inputInitController(const s32 cidx, const s32 jidx)
+		}
+	}
+	padsCfg[cidx].gyroSensorActive = sensorActive;
+#endif
```
Whitespace.

fgsfdsfgs commented on port/src/input.c:
```diff
@@ -361,6 +461,10 @@ static inline void inputCloseController(const s32 cidx)
 	pads[cidx] = NULL;
 	padsCfg[cidx].rumbleOn = 0;
 
+    // Zero out gyro deltas and orientation to prevent drift after disconnect
```
Whitespace.

fgsfdsfgs commented on port/src/input.c:
```diff
@@ -1319,6 +1566,537 @@ void inputSetMouseLockMode(s32 lockmode)
+		if (dx) *dx = (s32)gyroDeltaYaw[cidx];
+		if (dy) *dy = (s32)gyroDeltaPitch[cidx];
+		if (dz) *dz = (s32)gyroDeltaRoll[cidx];
+}
```
Whitespace.

fgsfdsfgs commented on port/src/input.c:
```diff
@@ -1319,6 +1566,537 @@ void inputSetMouseLockMode(s32 lockmode)
+	padsCfg[cidx].gyroAimMode = mode;
+}
+
+void applyGyroAimMode(s32 cidx, f32* deltaX, f32* deltaY, f32* deltaZ)
+{
+	s32 gyroAimMode = inputGetGyroAimMode(cidx);
+	(void)gyroAimMode;
+}
```
What is the point of this function?

AL2009man replied:
`inputGetGyroAimMode` (Aim Mode) is meant to control between `GYRO_AIM_MODE_CAMERA` (Gyro Camera only, no crosshair), `GYRO_AIM_MODE_CROSSHAIR` (Gyro Crosshair only, that's the chosen default!) or `GYRO_AIM_MODE_BOTH`(both Crosshair and Camera). You can see the inner works within `bondmove.c`

It's akin to how some major games (like Fortnite, The Finals and Call of Duty HQ) with gyro support does it: either "Always", "HipFire" or "ADS". but Perfect Dark has no concept of "ADS", only Crosshair.

However: this setup is based on [n64emu-SDL2GyroInjector](https://github.com/TauAkiou/n64emu-SDL2GyroInjector)'s Gyro Aim modes, but without the Wiimote-style aiming. (that's...probably be reserved for *after* this PR, i accidentally got it working but too buggy)

fgsfdsfgs replied:
No, I meant specifically `applyGyroAimMode`. It's only used in `input.c` and it's one line and doesn't use all of its arguments except `cidx`. Just inline it, it doesn't have to be a separate function.

AL2009man replied:
This has been...mostly addressed.

AL2009man replied:
> No, I meant specifically `applyGyroAimMode`. It's only used in `input.c` and it's one line and doesn't use all of its arguments except `cidx`. Just inline it, it doesn't have to be a separate function.

@fgsfdsfgs ok, this issue has been addressed...after having to fix regressions on the multiplayer side.

AL2009man replied:
ok, the rest of the Gyro processing has been made static, as per requested!

fgsfdsfgs commented on port/src/input.c:
```diff
@@ -1319,6 +1566,537 @@ void inputSetMouseLockMode(s32 lockmode)
+}
+
+void inputGyroGetScaledDelta(s32 cidx, f32* dx, f32* dy, f32* dz)
+{
+    if (!dx || !dy || !dz) return;
+
+    f32 gdx = 0.f, gdy = 0.f, gdz = 0.f;
+
+    if (padsCfg[cidx].gyroEnabled) {
+        if (!isnan(gyroDeltaYaw[cidx]) && !isnan(gyroDeltaPitch[cidx]) && !isnan(gyroDeltaRoll[cidx])) {
+            gdx = gyroDeltaYaw[cidx] * padsCfg[cidx].gyroSensX;
+            if (padsCfg[cidx].gyroInvertX) gdx = -gdx;
+            gdy = gyroDeltaPitch[cidx] * padsCfg[cidx].gyroSensY;
+            if (padsCfg[cidx].gyroInvertY) gdy = -gdy;
+            gdz = gyroDeltaRoll[cidx] * padsCfg[cidx].gyroSensY;
+        }
+    }
+
+    *dx = gdx;
+    *dy = gdy;
+    *dz = gdz;
+
+    applyGyroVHMixer(cidx, dx, dy);
```
Whitespace.

fgsfdsfgs commented on port/src/input.c:
```diff
@@ -1528,6 +2306,22 @@ PD_CONSTRUCTOR static void inputConfigInit(void)
+		configRegisterInt(strFmt("%s.gyroAimInvertY", secname), &padsCfg[c].gyroAimInvertY, 0, 1);
+		configRegisterFloat(strFmt("%s.gyroMinThreshold", secname), &padsCfg[c].gyroMinThreshold, 0.f, 1.f);
+		configRegisterFloat(strFmt("%s.gyroSmoothing", secname), &padsCfg[c].gyroSmoothing, 0.f, 1.f);
+		configRegisterInt(strFmt("%s.gyroAutoCalibration", secname), &padsCfg[c].gyroAutoCalibration, 0, 1);
```
Adding this many config options per player will probably overflow `CONFIG_MAX_SETTINGS`. Might need to increase it.
Probably capitalize all the setting names (i.e. `Gyro` instead of `gyro`).
Would generally be a good thing to consider which settings are actually *required*.

AL2009man replied:
the CONFIG_MAX_SETTINGS is at 300 right now, but I'll increase it to...500 for now

fgsfdsfgs commented on port/src/optionsmenu.c:
```diff
@@ -572,6 +572,452 @@ static MenuItemHandlerResult menuhandlerSwapSticks(s32 operation, struct menuite
+		     0,
+		     0,
+		     0,
+		     0,
```
Whitespace.

fgsfdsfgs commented on port/src/optionsmenu.c:
```diff
@@ -1627,27 +2081,30 @@ struct menubind {
-	{ CK_2000,   "Full Crouch [+]\n",   "N64 Ext 2000\n" },
-	{ CK_ACCEPT, "UI Accept [+]\n",     "EXT UI Accept\n" },
-	{ CK_CANCEL, "UI Cancel [+]\n",     "EXT UI Cancel\n" },
+		{ CK_ZTRIG,  "Fire [ZT]\n",         "N64 Z Trigger\n" },
+		{ CK_LTRIG,  "Fire Mode [LT]\n",    "N64 L Trigger\n"},
+		{ CK_RTRIG,  "Aim Mode [RT]\n",     "N64 R Trigger\n" },
+		{ CK_A,      "Use / Accept [A]\n",  "N64 A Button\n" },
+		{ CK_B,      "Use / Cancel [B]\n",  "N64 B Button\n" },
+		{ CK_START,  "Pause Menu [ST]\n",   "N64 Start\n" },
+		{ CK_DPAD_U, "D-Pad Up [DU]\n",     "N64 D-Pad Up\n" },
+		{ CK_DPAD_R, "D-Pad Right [DR]\n",  "N64 D-Pad Right\n" },
+		{ CK_DPAD_L, "Prev Weapon [DL]\n",  "N64 D-Pad Left\n" },
+		{ CK_DPAD_D, "Radial Menu [DD]\n",  "N64 D-Pad Down\n" },
+		{ CK_C_U,    "Forward [CU]\n",      "N64 C-Up\n" },
+		{ CK_C_D,    "Backward [CD]\n",     "N64 C-Down\n" },
+		{ CK_C_R,    "Strafe Right [CR]\n", "N64 C-Right\n" },
+		{ CK_C_L,    "Strafe Left [CL]\n",  "N64 C-Left\n" },
+		{ CK_X,      "Reload [X]\n",        "N64 Ext X\n" },
+		{ CK_Y,      "Next Weapon [Y]\n",   "N64 Ext Y\n" },
+		{ CK_8000,   "Cycle Crouch [+]\n",  "N64 Ext 8000\n" },
+		{ CK_4000,   "Half Crouch [+]\n",   "N64 Ext 4000\n" },
+		{ CK_2000,   "Full Crouch [+]\n",   "N64 Ext 2000\n" },
+		{ CK_RESET_CAMERA, "Reset Camera/Crosshair [+]\n", "EXT Reset Camera/Crosshair\n" },
+		{ CK_GYRO_MOD, "Gyro Modifier [+]\n", "EXT Gyro Modifier\n" },
+		{ CK_GYRO_CALIB, "Gyro Calibration (Manual) [+]\n", "EXT Gyro Calibration (Manual)[+]\n" },
+		{ CK_ACCEPT, "UI Accept [+]\n",     "EXT UI Accept\n" },
+		{ CK_CANCEL, "UI Cancel [+]\n",     "EXT UI Cancel\n" },
```
Whitespace.

fgsfdsfgs commented on src/game/bondmove.c:
```diff
@@ -43,6 +43,15 @@
+// Provide a fallback implementation if needed
+#ifndef fmax
+#define fmax(a, b) ((a) > (b) ? (a) : (b))
+#endif
```
Use `MAX()` if it's defined here. Don't remember for sure if it should be.

fgsfdsfgs commented on src/game/bondmove.c:
```diff
@@ -1412,34 +1487,35 @@ void bmoveProcessInput(bool allowc1x, bool allowc1y, bool allowc1buttons, bool i
-						g_Vars.currentplayer->swivelpos[0] = 0.f;
-						g_Vars.currentplayer->swivelpos[1] = 0.f;
-					}
+                    // Handle turning and looking up/down via mouselook or gyro when aiming
+                    bool allowcross = allowmcross || allowgcross;
+                    if (g_Vars.currentplayer->insightaimmode && allowcross && bgunGetWeaponNum(HAND_RIGHT) != WEAPON_HORIZONSCANNER) {
+                        if (g_Vars.currentplayer->swivelpos[0] > 0.9f) {
+                            movedata.aimturnrightspeed = (g_Vars.currentplayer->swivelpos[0] - 0.9f) / 0.1f;
+                            movedata.aimturnleftspeed = 0.f;
+                        } else if (g_Vars.currentplayer->swivelpos[0] < -0.9f) {
+                            movedata.aimturnleftspeed = (g_Vars.currentplayer->swivelpos[0] - -0.9f) / -0.1f;
+                            movedata.aimturnrightspeed = 0.f;
+                        }
+                        f32 vertaup = 0.f, vertadown = 0.f;
+                        if (g_Vars.currentplayer->swivelpos[1] > 0.9f) {
+                            vertaup = (g_Vars.currentplayer->swivelpos[1] - 0.9f) / 0.1f;
+                        } else if (g_Vars.currentplayer->swivelpos[1] < -0.9f) {
+                            vertadown = (g_Vars.currentplayer->swivelpos[1] - -0.9f) / -0.1f;
+                        }
+                        // Uninvert pitch if needed
+                        if (movedata.invertpitch) {
+                            movedata.speedvertaup = vertadown;
+                            movedata.speedvertadown = vertaup;
+                        } else {
+                            movedata.speedvertaup = vertaup;
+                            movedata.speedvertadown = vertadown;
+                        }
+                    } else {
+                        // Reset mouse/gyro aim position when not aiming
+                        g_Vars.currentplayer->swivelpos[0] = 0.f;
+                        g_Vars.currentplayer->swivelpos[1] = 0.f;
+                    }
```
Whitespace.

fgsfdsfgs commented on src/game/bondmove.c:
```diff
@@ -1946,10 +2022,34 @@ void bmoveProcessInput(bool allowc1x, bool allowc1y, bool allowc1buttons, bool i
-		offbike = g_Vars.currentplayer->bondmovemode == MOVEMODE_WALK
-			|| g_Vars.currentplayer->bondmovemode == MOVEMODE_GRAB;
+#ifndef PLATFORM_N64
+			// Camera recenter vertical binding
+			static bool prev_reset_pressed = false;
+			bool reset_pressed = inputResetCameraPressed(cidx);
+
+			if (reset_pressed && !prev_reset_pressed) {
+				if (g_Vars.currentplayer) {
+					if (g_Vars.currentplayer->insightaimmode) {
+						// Recenter crosshair if in crosshair aim mode
+						g_Vars.currentplayer->swivelpos[0] = 0.0f;
+						g_Vars.currentplayer->swivelpos[1] = 0.0f;
+					}
+
+					if (offbike && !g_Vars.currentplayer->insightaimmode) {
+						g_Vars.currentplayer->cachedlookahead = bmoveCalculateLookahead();
+						g_Vars.currentplayer->docentreupdown = true;
+						g_Vars.currentplayer->automovecentre = false;
+						g_Vars.currentplayer->speedverta = (g_Vars.currentplayer->vv_verta < g_Vars.currentplayer->cachedlookahead) ? 7.0f : -7.0f;
+						s_recenterfast = 1;
+					}
+				}
+			}
+			prev_reset_pressed = reset_pressed;
```
Maybe whitespace? Also, don't use `snake_case` in the game code, unless it's for a prefix like `g_`.

fgsfdsfgs commented on src/game/bondmove.c:
```diff
@@ -2068,7 +2169,15 @@ void bmoveProcessInput(bool allowc1x, bool allowc1y, bool allowc1buttons, bool i
+    }
+#else
+    g_Vars.currentplayer->vv_verta += g_Vars.currentplayer->speedverta * g_Vars.lvupdate60freal * 3.5f;
+#endif
```
Whitespace.

fgsfdsfgs commented on src/game/bondmove.c:
```diff
@@ -2204,23 +2315,50 @@ void bmoveProcessInput(bool allowc1x, bool allowc1y, bool allowc1buttons, bool i
-		}
+			if (allowgcross && g_Vars.currentplayernum == 0) {
+					// Gyro is active, apply gyro movement FIRST
+					inputGyroGetScaledDeltaCrosshair(g_Vars.currentplayernum, &movedata.gyrolookdx, &movedata.gyrolookdy);
+
+					if (movedata.gyrolookdx != 0.0f || movedata.gyrolookdy != 0.0f) {
+							const float xcoeff = 320.f / 1080.f;
+							const float ycoeff = 240.f / 1080.f;
+							const float xscale = (PLAYER_EXTCFG().gyroaimsensx * xcoeff * g_Vars.lvupdate60freal) / g_Vars.currentplayer->aspect;
+							const float yscale = PLAYER_EXTCFG().gyroaimsensy * ycoeff * g_Vars.lvupdate60freal;
+
+							float x = g_Vars.currentplayer->swivelpos[0] + movedata.gyrolookdx * xscale;
+							float y = g_Vars.currentplayer->swivelpos[1] + movedata.gyrolookdy * yscale;
+
+							g_Vars.currentplayer->swivelpos[0] = x;
+							g_Vars.currentplayer->swivelpos[1] = y;
+
+							return;
+					}
+			}
+
+			// Mouse input (only if gyro input was inactive)
+			if (allowmcross) {
+					const f32 xcoeff = 320.f / 1080.f;
+					const f32 ycoeff = 240.f / 1080.f;
+					const f32 xscale = (PLAYER_EXTCFG().mouseaimspeedx * xcoeff) / g_Vars.currentplayer->aspect;
+					const f32 yscale = PLAYER_EXTCFG().mouseaimspeedy * ycoeff;
+
+					f32 x = g_Vars.currentplayer->swivelpos[0] + movedata.freelookdx * xscale;
+					f32 y = g_Vars.currentplayer->swivelpos[1] + movedata.freelookdy * yscale;
+
+					x = (x < -1.f) ? -1.f : ((x > 1.f) ? 1.f : x);
+					y = (y < -1.f) ? -1.f : ((y > 1.f) ? 1.f : y);
+
+					g_Vars.currentplayer->swivelpos[0] = x;
+					g_Vars.currentplayer->swivelpos[1] = y;
+					bgunSwivelWithDamp(x, y, 0.01f);
+
+					return;
+			}
```
Whitespace. Also maybe consider separating this out into a function that takes `aimspeedx/y` and `dx`/`dy` to avoid duplicating code for the mouse and gyro.

AL2009man replied:
we'll reserve that for a separate PR, after Gyro Aiming support gets merged. I figured out how to fix crosshair sensitivity being tied to framerate, and I believe it's better to do two birds-one-stone. 

in the meantime: a temporary workaround will be added in across both this PR and https://github.com/fgsfdsfgs/perfect_dark/pull/620

AL2009man replied:
@fgsfdsfgs quick update: the crosshair scaling/clamping logic has been separated into its own `bmoveApplyCrosshairMovement` centralized system, to prevent duplicating code. This has been applied to both this PR and https://github.com/fgsfdsfgs/perfect_dark/pull/620

fgsfdsfgs commented on src/game/bondmove.c:
```diff
@@ -2204,23 +2315,50 @@ void bmoveProcessInput(bool allowc1x, bool allowc1y, bool allowc1buttons, bool i
+
+			// Default joystick-based movement if neither mouse nor gyro crosshair movement is active
+			bgunSwivelWithoutDamp((movedata.c1stickxraw * 0.65f) / 80.0f,
+					(movedata.c1stickyraw * 0.65f) / 80.0f);
```
Whitespace.

fgsfdsfgs commented on src/game/player.c:
```diff
@@ -3678,6 +3678,49 @@ void playerTick(bool arg0)
+								g_Vars.currentplayer->swivelpos[0] += gdx_crosshair; // Crosshair movement
+								g_Vars.currentplayer->swivelpos[1] += gdy_crosshair;
+						}
+				}
```
Whitespace.

fgsfdsfgs commented on src/include/game/bondmove.h:
```diff
@@ -29,7 +29,7 @@ f32 bmoveGetSpeedThetaControlLimit(f32 value);
 f32 bmoveCalculateLookahead(void);
 void bmoveResetMoveData(struct movedata *data);
-void bmoveProcessInput(bool allowc1x, bool allowc1y, bool allowc1buttons, bool ignorec2);
+void bmoveProcessInput(s32 cidx, bool allowc1x, bool allowc1y, bool allowc1buttons, bool ignorec2);
```
There's no reason to add this argument here, it's literally just `g_Vars.currentplayernum`. If you want a shorter var name to use, make it a local `const s32` in the function.

AL2009man replied:
please check. right now: it's sorta a hacky workaround to also fix Gyro-related functions as a result of this removal.

fgsfdsfgs commented on port/gamepadmotionhelper/GamepadMotion.h:
```diff
@@ -0,0 +1,99 @@
+void SetStillnessErrorClimbRate(GamepadMotionHandle handle, float value);
+float GetStillnessErrorClimbRate(GamepadMotionHandle handle);
+void SetStillnessErrorDropOnRecalibrate(GamepadMotionHandle handle, float value);
+float GetStillnessErrorDropOnRecalibrate(GamepadMotionHandle handle);
+void SetStillnessCalibrationEaseInTime(GamepadMotionHandle handle, float value);
+float GetStillnessCalibrationEaseInTime(GamepadMotionHandle handle);
+void SetStillnessCalibrationHalfTime(GamepadMotionHandle handle, float value);
+float GetStillnessCalibrationHalfTime(GamepadMotionHandle handle);
+void SetStillnessConfidenceRate(GamepadMotionHandle handle, float value);
+float GetStillnessConfidenceRate(GamepadMotionHandle handle);
+void SetStillnessGyroDelta(GamepadMotionHandle handle, float value);
+float GetStillnessGyroDelta(GamepadMotionHandle handle);
+void SetStillnessAccelDelta(GamepadMotionHandle handle, float value);
+float GetStillnessAccelDelta(GamepadMotionHandle handle);
+void SetSensorFusionCalibrationSmoothingStrength(GamepadMotionHandle handle, float value);
+float GetSensorFusionCalibrationSmoothingStrength(GamepadMotionHandle handle);
+void SetSensorFusionAngularAccelerationThreshold(GamepadMotionHandle handle, float value);
+float GetSensorFusionAngularAccelerationThreshold(GamepadMotionHandle handle);
+void SetSensorFusionCalibrationEaseInTime(GamepadMotionHandle handle, float value);
+float GetSensorFusionCalibrationEaseInTime(GamepadMotionHandle handle);
+void SetSensorFusionCalibrationHalfTime(GamepadMotionHandle handle, float value);
+float GetSensorFusionCalibrationHalfTime(GamepadMotionHandle handle);
+void SetSensorFusionConfidenceRate(GamepadMotionHandle handle, float value);
+float GetSensorFusionConfidenceRate(GamepadMotionHandle handle);
+void SetGravityCorrectionShakinessMaxThreshold(GamepadMotionHandle handle, float value);
+float GetGravityCorrectionShakinessMaxThreshold(GamepadMotionHandle handle);
+void SetGravityCorrectionShakinessMinThreshold(GamepadMotionHandle handle, float value);
+float GetGravityCorrectionShakinessMinThreshold(GamepadMotionHandle handle);
+void SetGravityCorrectionStillSpeed(GamepadMotionHandle handle, float value);
+float GetGravityCorrectionStillSpeed(GamepadMotionHandle handle);
+void SetGravityCorrectionShakySpeed(GamepadMotionHandle handle, float value);
+float GetGravityCorrectionShakySpeed(GamepadMotionHandle handle);
+void SetGravityCorrectionGyroFactor(GamepadMotionHandle handle, float value);
+float GetGravityCorrectionGyroFactor(GamepadMotionHandle handle);
+void SetGravityCorrectionGyroMinThreshold(GamepadMotionHandle handle, float value);
+float GetGravityCorrectionGyroMinThreshold(GamepadMotionHandle handle);
+void SetGravityCorrectionGyroMaxThreshold(GamepadMotionHandle handle, float value);
+float GetGravityCorrectionGyroMaxThreshold(GamepadMotionHandle handle);
+void SetGravityCorrectionMinimumSpeed(GamepadMotionHandle handle, float value);
+float GetGravityCorrectionMinimumSpeed(GamepadMotionHandle handle);
```
If this C wrapper was made by you, consider adding a prefix to the names of these functions to more easily differentiate them. Like gm or something.

AL2009man replied:
I've revised the entire C Wrapper functionality alongside GamepadMotion itself (now closer to the original forked version). To make things easier to manage (in case of future GMH Updates): I added a define function to cover the prefix portion. 

the header wrapper looked to have duplicated code, but it should not affect the real wrapper portion.

AL2009man committed:
cleaned up as many remaining whitespaces

AL2009man committed:
replaced direct names with numbered entries instead.

this avoids issues with the 32-bit top buttons breakage.

AL2009man committed:
added define feedback, and swapped the way how camera recenter works

AL2009man committed:
Simplified Gyro Y/X Invert toggle

AL2009man committed:
cleaned up and reorganzied `GyroGetSpeed` to mirror GetMouseSpeed

this also remove leftover code and redundancies

AL2009man committed:
cleaned up Gyro Deadzone slider internally

AL2009man committed:
cleaned up GyroAxisMode

AL2009man committed:
workarounding `bmoveProcessInput` to remove `s32 cidx`

this is a hacky workaround due to existing Gyro-related code that needs to be revised a bit.

AL2009man committed:
reorganized gyro api on input.h

AL2009man committed:
increased `CONFIG_MAX_SETTINGS` to `500`

AL2009man committed:
reorganized crosshair movement

to avoid duplications between gyro and mouse crosshairs, the actual crosshair function has been decoupled.

AL2009man committed:
added prefix to GamepadMotionHelper-related code, capitalize gyro settings on pd.ini

AL2009man committed:
Merge branch 'fgsfdsfgs:port' into Gyro-Aim

AL2009man committed:
Gyro Crosshair is no longer tied to framerate, restored old crosshair setup

Gyro Crosshair should now behave consistently across all framerates. Gowever: the unified crosshair setup will be reverted back to it's original setup for the time being. best to keep it for a separate PR where we can finally solve the crosshair framerate problem once and for all.

AL2009man committed:
slightly tweaked Gyro Sensitivity Scaling, also use FOV scaling

it should improve 360 degree horizontal rotation a bit.

AL2009man committed:
reorganize camera reset

AL2009man committed:
Camera/Crosshair Recentering smooted out.

this commit improves the way how Camera/Crosshair transition are handled. instead of instant: it should now be smooth, while still retaining control after the animation state ends.

AL2009man committed:
renamed s_recenterfast to s_recentercamera, now that we have a proper smoothed out animation

AL2009man committed:
fixed Turn/Look Up+Down while using Crosshair Aiming

You can now look up and down (joystick) while using Gyro/Mouse Input's Crosshair Aiming mode

AL2009man committed:
editorized the joystick turning/looking note

AL2009man committed:
one last minute editorize name to be more generic

AL2009man committed:
Delete .vscode directory

...this shouldn't be there

AL2009man committed:
reverted joystick turn aim

will instead moving it to a separate PR

AL2009man committed:
added Gyro Tightening slider.

Added Gyro Tightening slider, which reduce sensitivity during smaller movement threshold before ramping up back to normal levels. This should help with shaky hands, best combined with Gyro Smoothing Threshold

also: renamed internal old GyroMinTreshold with Gyro Deadzones to be more accurate about it,

AL2009man committed:
simplified Smoothing, Tighting and Deadzone sliders internally

AL2009man committed:
fixed Gyro X/Y Output Mixer slider

for some reason: the slider can jump by 2-3 percent, but that should be fixed. it should be easier to use the same 16:9 ratio (-43%) as old SteamInput gyro.

AL2009man committed:
cleaned up GyroAxisMapping

AL2009man committed:
added additional credits

credited GyroWiki and SDL2GyroInjector

AL2009man committed:
Gyro Calibration UI added to Gyro Menu

Added the ability to manually calibrate the controller's Motion Sensors within the Menu UI, no need to press the Calibration button! Just place the controller on a surface, and wait for 5 seconds, giving you ample time to put the controller on the floor, as the calibration is instant! 

If you accidentally move the controller during the gyro calibration process: press the ACCEPT key two times inorder to restart the Gyro Calibration process!

AL2009man committed:
fixed regression with Gyro Calibration

AL2009man committed:
added `inputPauseGyro` to streamline gyro input leak

to make things easier to handle cases like Window Focus or reconnections: a dedicated static function was added.

AL2009man committed:
fixed Gyro Input connection handlingly

ditched the custom Input Handle infavor `inputInitController` setup.

Do note that reconnecting the controller will make horizontal movement almost none before recovering.

AL2009man committed:
Improved Gyro AutoCalibration, fixed Reset Camera/Crosshair function, improved Gyro Filters

Completely revised the internal Gyro Calibration handle to be easier to organize and read, AutoCalibration has been improved to account for Gyro Noise Threshold.

Additionally, a selection of improvements has been made to other gyro features and non-gyro features:

* Natural Sensitivity Scaling has been revised for the last time(?), by replacing `frameScale`
* Gyro Deadzone, Tightening (now based on Soft Tiered Smoothing) and Smoothing has been revised
* Camera/Crosshair Recenter has been improved, now it should center to horizon properly.
* removing the remaining bloat

AL2009man committed:
reorganized inputUpdateGyro and reintroduce Sensor Data Rate

portions of inputUpdateGyro has been separated for easier management, while reintroducting Sensor Rate

AL2009man committed:
integrate more GamepadMotionHelper functions,, Improved Controller Hotswap, redo the handles

more GamepadMotionHelper functions are being ultilized, no need to include hacky workarounds for this one. this also includes improvements to the Gyro Autocalibration status while renaming a few structs

AL2009man committed:
improved autocalibration status

it should have a proper cooldown timer. if done correctly: it'll take 5-6 seconds to do a complete auto-calibration after the controller is placed on the ground.

AL2009man committed:
added additional safetynet if the Gyro Autocalibration is still on the ground

AL2009man committed:
replaced Camera/Crosshair Recenter funciton to direct function

AL2009man committed:
Increarsed Gyro Autocalibration detection rate

Gyro Noise Threshold has been improved, it has gotten to the point that you can't stay still, bypassing the gyro autocalibration systems while still holding the controller.

additionally: portions of GamepadMotionSettings has been integrated, further replacing custom timing solutions.

AL2009man committed:
updated GamepadMotionHelper C wrapper to be more abstracted

AL2009man committed:
applyGyroAimMode "inlined", fixed Gyro orientation change after switching to other Axis spaces

AL2009man committed:
tweaked Gyro Space orientation

AL2009man committed:
organized Gyro Calibration handle

noticed there's duplicated GamepadMotionSetting stuffs, that's been moved over into a singular handle

AL2009man committed:
fixed GamepadMotionSettings applying to Manual instead of Auto-Calibration

AL2009man committed:
Gyro Autocalibration is now enabled by default

this is the final test stretch, also one last name change (will be locked in)

AL2009man committed:
fixed regression with Camera/Crosshair Recentering in Multiplayer

AL2009man committed:
fixed regression on Gyro Crosshair in Multiplayer

AL2009man committed:
blocked CK_0100/Gyro Manual Calibrate button during Gyro Calib UI process

AL2009man committed:
fixed regression with `gyroEnable` function

Gyro settings should be properly disabled and fallback to traditional joystick aiming  if using a Controller that lacks Motion Sensors (like Xbox controller)

AL2009man committed:
Enhanced Motion Sensor detection system

implemented `inputControllerHasMotionSensors` to better handle detection rate.

It's more evident within Controller Settings: where "Gyro Settings..." becomes inaccessible when using a unsupported controller (like Xbox Controller) or input remapper that lacks specific controller type's emulation.

AL2009man committed:
Cleaned up documentation for gyro auto-cal constants

AL2009man committed:
`GyroCalibrationHandle` now account for `inputControllerMotionSensorsSupported`

To ensure Gyro Calibration doesn't leak to non-gyro controller: `inputControllerMotionSensorsSupported` will be ultilized.

AL2009man committed:
added "Gyro Calibration (Manual)" action to defaulted N64 bindings

In case a player has a third-party Nintendo 64 Controller with Motion Sensor support, N64-style preset now has `MISC1` and `F10` functions by default

AL2009man committed:
remove dead code leftovers

AL2009man committed:
Adjusted Gyro Smoothing and Deadzone slider

Gyro Smoothing has been reduced oevrall to improve responsiveness and Deadzone should be smoothed out. Because of that: the default slider has been tweaked to better reflect that + Gyro Autocalibration still being enabled as default

AL2009man committed:
tweaked GyroAxis Switch

AL2009man committed:
Revised GamepadMotionHelper

to make things significantly easier to maintain, GamepadMotionHelper code will now be closer to the original fork version, while added a define prefix alias for easier management.

AL2009man committed:
improved GamepadMotionHelper compiler compatibility

AL2009man commented:
> that being out of the way, I will be requesting User Testing for this. This PR will be ready for review soon.
> 
> You can download it via GitHub Actions, and see how this works. For the best results: get a PlayStation 4/5/Nintendo Switch controller.
> 
> What needs to be tested:
> 
> * Local Multiplayer
> * HoverBike and Gadgets (Gyro Aim is disabled during SpyCam, which is fine within the context)
> * Gyro Autocalibration (right now: it's disabled by default)
>   ~- Update README.md to include Gyro-related information~
> * picking some of the best defaults based on user feedback
> 
> Known Issues: ~- Gyro Crosshair sens being tied to framerate (can't be fixed for now, seems to be universal issue)~ ~- Gyro Smoothing doesn't truly kicks in when the slider is above 0.60~
> 
> * Natural Sensitivity Scale may not function properly on Nintendo Switch controllers (already addressed in a future SDL2/3 release)

as a reminder: I'm still looking for people who can test Gyro Aim support. I already went thru some minor revisions and bug fixes, it should be a bit better. default gyro settings has been revised a bit

* Gyro Autocalibration is now enabled by default
* Gyro Smoothing and Deadzones has been upgraded. it should now be more responsive.
* attempts to reduce the rate of dropped sensitivity scaling 


the Gyro Autocalibration portion is where I think it covers edge-cases, so it's now enabled by default, to reflect that: the Smoothing, Tightening and Deadzones has been tweaked to reflect the changes.

and with that: I'll gonna need help on testing OS Platforms, as MacOS and Linux builds hasn't been tested yet. I can plan on testitng it on Linux via Steam Deck, but I don't have access to Mac device

AL2009man committed:
changed Gyro processing function to be static

AL2009man committed:
fixed Crosshair Sway detection when using Gyro/Mouse and Joystick

This fixes a issue where Joystick Sway conflicts with Mouse and Gyro's own crosshair swaying.

This is achieved by creating a entire input-driven detection system where the crosshair sway reduction "deadzone" will be switched based on active input.

AL2009man committed:
fixed Gyro AutoCalibration snap-back effect

This is more of a workaround than a fix, where there's a small chance that the confidence system may cause a "snap-back". This will use a Gyro pause mechanic to prevent it

AL2009man committed:
fixed bug with Gyro Autocalibration still active during Gyro Calibrartion UI

AL2009man committed:
adjusted comment notes to only the essentials, adjusted `GYRO_NOISE_THRESHOLD` to 0.2f

AL2009man committed:
added `InputWindowFocus` to address gyro focus loss

When the Window focus is lost: Gyro Input will continue to move ittself. This introduces a new flag where all Inputs (including Gyro) will be pause whenever the Window focus is lost.

This function can be disabled within pd.ini

AL2009man committed:
document hotfix

forgot to move that comment note to `inputHasWindowFocus`

AL2009man committed:
organized and reduce duplicated code

decided to clean up and try to centralize minor parts of Gyro Calibration.

AL2009man commented:
@fgsfdsfgs i've been doing final preparations (and hopefully addressing nearly all the issues raised and final batch of organization change to prevent the "overcomplication code" as best as I can) to prepare for another review request.

While I've tested it on both Windows and Linux (via Steam Deck): I don't have access to MacOS nor Nintendo Switch console (for homebrew), so I might need assistance on that front.

that being said: due to commit history being bloated: I suggest using the "Squash" merge if you plan to merge it on this PR. I got no problem making a "duplicated PR" to avoid commit bloat if needed.

AL2009man committed:
Organized Motion Sensor data

AL2009man committed:
reduced Gyro Calibration bloatness

AL2009man committed:
fixed regression with GyroCalibration UI

AL2009man committed:
semi-improved Gyro Auto-Calibration

cleaned up Gyro AutoCalibration process for abit, alongside additional notes to better explain the process, with some sysLogPrintf (it might be removed in the next commit?)

AL2009man committed:
Organized the entire Gyro Calibration code

reorganized Gyro Calibration handle to make it easier to actually read and organize

AL2009man committed:
tweaked gyro noise filter for Gyro Autocalibration

AL2009man committed:
Gravity Vector is now used for `GYRO_WORLD`

AL2009man committed:
Tighten up Gyro AutoCalibration detection rate

AL2009man committed:
updated README.md to better match the rest of the formats

AL2009man committed:
Merge branch 'fgsfdsfgs:port' into Gyro-Aim

AL2009man committed:
Centralized Camera/Crosshair Recenter and Scaling system

centralized both Crosshair scaling logic, and Camera/Crosshair Recentering logic to address duplicated issues, while fixing Camera Recenter when using Hoverboard

AL2009man committed:
remove leftover notes

AL2009man committed:
remove some of the LogPrintfsince it's no longer needed

AL2009man committed:
fixed regression on LogPrint for Gyro Autocalibration enable/disable

AL2009man committed:
Additional Gyro Auto-Calibration options, Cleaned up overall Gyro code

Added 3 new ways to Auto-Calibrate your Gyro, inspirited by Fortnite's new Calibration option

- Disabled: Will disable Gyro Autocalibration
- While In Menu: Will Auto Calibrate the Gyro only in a Menu dialog.
   -  this reuses the existing Gyro Autocalibration feature as before, same timer too!
   - This is now the default.
- While Stationary: Will autocalibrate Gyro when the Controller is placed on a flat surface/stationary
- Always: will try to autocalibrate the Controller whenever possible, will be more drift-prone.

Additionally: "While in Menus" and "Disabled" now shares the same Gyro Calibration saved offset, to better match how Fortnite handles their calibration setup.

AL2009man committed:
whitespace fix on `inputInitController`

AL2009man committed:
fixed regression on Gyro Calibration button (`CK_0100`)

at some point during the cleanup: a bug related to Gyro Calibration got slipped thru. This should be fixed.

AL2009man committed:
last minute note change on Gyro Calibration system

AL2009man commented:
hold on, i'm putting this PR on hold as i'm still ironing out.

taht being said: there's a likely chance that Gyro Aiming PR will have be moved over to a new repository to avoid large commit history, as i have no idea if all of this will be regular merge, or squash

AL2009man committed:
Cleaned up Gyro Calibration systems,

in a effort to minimize the amount of bloat whenever possible: the entire system has been cleaned up, with improved detection rate during AutoCalibration.
